### DISCUSSION TITLE
dumpViews - progress option not working?

### DISCUSSION THREAD

**Discussion by frnora** [clientIO/joint#2695] [Q&A] [answered]
### Introduction

Hi,

I began updating our app to v4. One of the first issues I ran into is that I cannot center the `paperScroller` after loading data anymore without having called `dumpViews` (the box reports 0 width and height otherwise). The paper is async.

Long story short:
Looking at [dumpView](https://docs.jointjs.com/api/dia/Paper/#dumpviews) it says it accepts a `progress` key. Does this actually work? Or am I doing something incorrectly?
`app.paper.dumpViews({ progress: function() { debugger; } });` <- should this pause on the `debugger` statement?

Thanks

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS, JointJS+

**Comment by kumilingus** [comment, marked as answer]
It's a mistake. There is no `progress` for `dumpViews`. It makes no sense as `dumpViews` is always synchronous.

Are you saying it used to work in `v3` and not in `v4`? There were no changes in this aspect as far as I know.

You don't need to force the paper content to rendered in order to center the paper scroller. You can use `useModelGeometry: true ` instead.

```ts
paperScroller.centerContent({ useModelGeometry: true });
```

It's missing from the docs right now. I am sorry for the inconveniences.
It's the same option as `getContentArea()` [useModelGeometry](https://docs.jointjs.com/api/dia/Paper/#getcontentarea) option.

**Comment by frnora** [comment]
I was actually thinking whether it is not a mistake, thinking `dumpViews` shouldn't be async. Thanks for confirming.

The setup is a bit more complicated because we have some workarounds in place for the Safari bug not rending the arrowheads, so we are waiting for the next animation frame, etc. The (most likely) issue is that we synchronously change the default router which calls `freeze/unfreeze` on the paper right after calling `fromJSON`. But... yes, it worked in v3 and now in v4 it doesn't, at least not without `dumpViews`. Very same setup.

The `useModelGeometry: true` sounds much better indeed, we will change to that.

Thanks a lot for your help!

On an (un)related note - I saw your comment that you fixed that Safari bug in 4.0.4 - these "rerendering" workarounds won't be needed anymore? (Now that would be great, I would love to drop it!)

**Reply by kumilingus** [reply]
*JointJS* `4.0.4` has been released and it can be used with *JointJS+*`4.0.1` already. Please see https://github.com/clientIO/joint/discussions/2652.

**Reply by frnora** [reply]
~Is there no bundled version of _JointJS_ anymore? The only bundle that exists is in _JointJS+_?~ EDIT: got it

We are using the _+version_ bundle, we are not building it. The only thing we build is our application, so no import statements from _joint_ (taking it from global). Any easier way (than adding the dependencies and changing the imports in our app) to test it?
